### PR TITLE
Vecsim add vector check

### PR DESCRIPTION
**Pull request by GuyAv46**
added a validation check at insertion to vector index

### CODE REVIEW

File: src/spec.c [hunk 1]
```diff
@@ -539,18 +539,21 @@ static int parseVectorField(FieldSpec *fs, ArgsCursor *ac, QueryError *status) {
...
+    fs->expBlobSize = fs->vecSimParams.bfParams.dim * VecSimType_sizeof(fs->vecSimParams.bfParams.type);
   } else if (!strncasecmp(VECSIM_ALGORITHM_HNSW, algStr, len)) {
     fs->vecSimParams.algo = VecSimAlgo_HNSWLIB;
     fs->vecSimParams.hnswParams.initialCapacity = 1000;
     fs->vecSimParams.hnswParams.M = HNSW_DEFAULT_M;
     fs->vecSimParams.hnswParams.efConstruction = HNSW_DEFAULT_EF_C;
     fs->vecSimParams.hnswParams.efRuntime = HNSW_DEFAULT_EF_RT;
-    return parseVectorField_hnsw(fs, ac, status);
+    rc = parseVectorField_hnsw(fs, ac, status);
+    fs->expBlobSize = fs->vecSimParams.hnswParams.dim * VecSimType_sizeof(fs->vecSimParams.hnswParams.type);
```
Comment (DvirDukhan) [inline src/spec.c:551]: please have it on the parse function, and set it only upon successful parse, so if the parse fails you will not set/read invalid data

**Review by DvirDukhan — APPROVED** [review]

File: tests/pytests/test_vecsim.py [hunk 2]
```diff
@@ -653,3 +653,21 @@ def test_single_entry(env):
...
+
+    conn.execute_command('FT.CREATE', 'idx', 'SCHEMA', 'v', 'VECTOR', 'HNSW', '6', 'TYPE', 'FLOAT32', 'DIM', dimension, 'DISTANCE_METRIC', 'L2')
+
+    vector = np.random.rand(1+dimension).astype(np.float32)
+    conn.execute_command('HSET', '3', 'v', vector[:dimension-1].tobytes())
+    conn.execute_command('HSET', '4', 'v', vector[:dimension].tobytes())
+    conn.execute_command('HSET', '5', 'v', vector[:dimension+1].tobytes())
+    
+    waitForIndex(env, 'idx')
+    assertInfoField(env, 'idx', 'num_docs', '2')
```
Comment (ashtul) [inline tests/pytests/test_vecsim.py:673]: add check for info field `hash_indexing_failures` == 4

File: src/vector_index.c [hunk 3]
```diff
@@ -69,11 +69,10 @@ VecSimIndex *OpenVectorIndex(RedisSearchCtx *ctx,
   return openVectorKeysDict(ctx, keyName, 1);
 }
 
-static bool isFit (VecSimIndex *ind, size_t size) {
+static size_t expBlobSize (VecSimIndex *ind) {
```
Comment (ashtul) [inline src/vector_index.c:72]: remove space to `(`

**Review by ashtul — COMMENTED** [review]
Look good.
Minor comments

File: src/field_spec.h [hunk 4]
```diff
@@ -92,6 +92,8 @@ typedef struct FieldSpec {
 
   // Vector similarity index parameters.
   VecSimParams vecSimParams;
+  // expected size of vector blob.
+  size_t expBlobSize;
```
Comment (alonre24) [inline src/field_spec.h:96]: Why adding a new field to the generic struct? `expBlobSize` can be deduced from `vecSimParams`

File: tests/pytests/test_vecsim.py [hunk 5]
```diff
@@ -653,3 +653,21 @@ def test_single_entry(env):
...
+    conn.execute_command('HSET', '0', 'v', vector[:dimension-1].tobytes())
+    conn.execute_command('HSET', '1', 'v', vector[:dimension].tobytes())
+    conn.execute_command('HSET', '2', 'v', vector[:dimension+1].tobytes())
+
+    conn.execute_command('FT.CREATE', 'idx', 'SCHEMA', 'v', 'VECTOR', 'HNSW', '6', 'TYPE', 'FLOAT32', 'DIM', dimension, 'DISTANCE_METRIC', 'L2')
+
+    vector = np.random.rand(1+dimension).astype(np.float32)
+    conn.execute_command('HSET', '3', 'v', vector[:dimension-1].tobytes())
+    conn.execute_command('HSET', '4', 'v', vector[:dimension].tobytes())
+    conn.execute_command('HSET', '5', 'v', vector[:dimension+1].tobytes())
```
Comment (alonre24) [inline tests/pytests/test_vecsim.py:670]: Consider checking the error message as well for ids `3` and `5`, as you did for `ft.search`

**Review by alonre24 — COMMENTED** [review]
Looks great! added 2 comments

File: src/field_spec.h (hunk 4 above)
Comment (GuyAv46) [reply src/field_spec.h:96]: yes, but do we want to calculate it on every insertion?

File: tests/pytests/test_vecsim.py (hunk 5 above)
Comment (GuyAv46) [reply tests/pytests/test_vecsim.py:670]: there is no error message, as `HSET` command is legal even if we can't index it.

File: src/field_spec.h (hunk 4 above)
Comment (ashtul) [reply src/field_spec.h:96]: Good point.
We can add it in `VecSimParams` as `int` after `algo` and with padding, not consume more memory. WDYT?

Comment (alonre24) [reply src/field_spec.h:96]: I see. I would adding `expBlobSize` to `vecSimParams` - but that will require changes in VecSim -  so maybe we should leave it for now...

Comment (GuyAv46) [reply src/field_spec.h:96]: I think it’s not a parameter for creating the index so it doesn’t need to be there, but maybe we can add to the FieldSpec a union? There are few things that relate only for specific indexing types

Comment (DvirDukhan) [reply src/field_spec.h:96]: this is not related to index creating so it should not be in `vecSimParams`. Same as tag field has 'tagSep', I think for now it is a valid solution

File: tests/pytests/test_vecsim.py (hunk 5 above)
Comment (alonre24) [reply tests/pytests/test_vecsim.py:670]: @ashtul Isn't there a way of retrieving error message in such case? `ft.info` only contains the number of failures right?

File: src/field_spec.h (hunk 4 above)
Comment (alonre24) [reply src/field_spec.h:96]: > I think it’s not a parameter for creating the index so it doesn’t need to be there, but maybe we can add to the FieldSpec a union? There are few things that relate only for specific indexing types like the tag only params

Agree, union it's a good idea for future refactor

File: tests/pytests/test_vecsim.py (hunk 5 above)
Comment (ashtul) [reply tests/pytests/test_vecsim.py:671]: @alonre24 No. there is no way to check it while inserting docs.
Instead, Guy added a test for `hash_indexing_failures` parameter

Comment (DvirDukhan) [reply tests/pytests/test_vecsim.py:671]: @GuyAv46 if there is no way to test the error message upon insertion, please remove it/log it with low severity

Comment (GuyAv46) [reply tests/pytests/test_vecsim.py:671]: OK

**Review by DvirDukhan — APPROVED** [review]

**Review by ashtul — APPROVED** [review]
LGTM

**Pull request merged**